AirSkyBoat/AirSkyBoat
# Fixes mobskill chip dmg and additional effects

**_I affirm:_**


## Please enter a player-facing description

Fixes an issue which was causing players to lose 1 hp per shadow lost from mob weapon skills which were blocked by shadows.
Also fixes numerous issues where a mob skill would apply its additional effect even if abosrobed by shadows.

## What does this pull request do? (Please be technical)

**Chip Damage through shadows**
Most mobskills adhere to the following call pattern 
_please note params which are not pertinent to this PR are snipped_

```
local info = xi.mobskills.mobPhysicalMove(...)
local dmg = xi.mobskills.mobFinalAdjustments(info.dmg, ..., info.hitslanded)
target:takeDamage(dmg)
return dmg
```
Here it is important to note four things for this PR.
1. info.hitslanded is passed in the "shadowbehavior" param for mobFinalAdjustments
2. internal to mobFinalAdjustments, the target having shadows is handled.  If the target has equal or greater shadows vs hitslanded, mobFinalAdjustments changes the message of the skill to xi.msg.basic.SHADOW_ABSORB and returns the number of shadows absrobed.
3. this number of shadows absorbed returned is assigned to dmg, and returned out of the lua layer.  Combined with the skill message set to xi.msg.basic.SHADOW_ABSORB, we achieve the correct client display of "X shadows absorb the attack".
4. Dmg now being equal to shadows absorbed means that target:takeDamage(dmg) will apply 1-4 dmg to the target

**Additional effects**
Some mob skills (like spike ball) also contained code like the following
```
local dmg = xi.mobskills.mobFinalAdjustments(info.dmg, ..., info.hitslanded)
target:takeDamage(dmg)

if dmg > 0 then
    mobStatusEffectMove(.....)
end
return dmg
```
In this scenario, when a move was fully absorbed by shadows, the dmg value was equal to shadows removed (not 0) and the status effect would be applied.

Ideally, physical moves should use mobPhysicalStatusMove - which has an internal gate to check if the skill has a miss message.

**Blind application of statuses**
Some skills did not even gate additional effects or hate reset behind checking dmg
In this scenario a player could evade a move or "blink" a hit via Utsu and still recieve the full additional effect of the miss.
I have adjusted skills which were applying statuses to honor actually hitting the player.

**Hate reset**
Some physical skills (deal_out, entangle, great_wheel, hoof_volley, hydro_shot, omega_javelin, sideswipe, vortex) reset hate and can miss/be absorbed by shadows.  These moves now require the player to get hit.
~~I have not modified how they work today, but need confirmation if they should have their hate reset gated behind actually hitting.  It sure seems like they should - but thats a relatively large change.~~
Thanks to Spaggi have some evidence that hate resets that dont actualy hit the player, dont reset hate.
[2 min mark - Nin loses two shadows, doesnt get "actually hit", mnk weaponskills, and nin keeps hate](https://www.youtube.com/watch?v=tYXS5uXRUU8)

**Future consideration**
I considered refactoring mobFinalAdjustment to provide multiple paramaters back to the caller (a table).
- dmg
- shadows absorbed
- actual hits against player after shadows

I did not take this pathway for three reasons.
1. this would require as many or more mobskill changes
2. this would push more responsibility onto each mob skill beyond the solution i chose (more "client" responsiblity if you think of the mobskill as the client and mobFinalAdjustments as a service)
3. We do not currently have a need for actual hits against the player after shadows to be exposed to the mob skill (but we will with [Tornado Edge](https://ffxiclopedia.fandom.com/wiki/Tornado_Edge))
4. (3a) Alternatively, Tornado Edge could be implemented as literally 3 distinct hits instead of a 3 hit mob skill. 
## Steps to test these changes

Spikeball is a good choice.
Verify that when the OG opressors of the galka (antica) use spikeball that 1) no poison is applied to moves blocked by shadows/blink and 2) the players HP is not reduced by 1 for each shadow removed (spikeball is single hit).


## Special Deployment Considerations

None
